File: 1399.count-largest-group.cpp
/*
 * @lc app=leetcode id=1399 lang=cpp
 *
 * [1399] Count Largest Group
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int countLargestGroup(int n) {
    std::unordered_map<int, int> counts;
    for (int i = 1; i <= n; i++) {
      int work = i;
      int sum = 0;
      while (work) {
        sum += work % 10;
        work /= 10;
      }
      counts[sum]++;
    }
    std::vector<int> vector_count;
    for (auto count : counts) {
      vector_count.push_back(count.second);
    }
    int max = *std::max_element(vector_count.begin(), vector_count.end());
    int ret = 0;
    for (int count : vector_count) {
      if (count == max) {
        ret++;
      }
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  ASSERT_EQ(s.countLargestGroup(1), 1);
  ASSERT_EQ(s.countLargestGroup(2), 2);
  ASSERT_EQ(s.countLargestGroup(9), 9);
  ASSERT_EQ(s.countLargestGroup(10), 1);
  ASSERT_EQ(s.countLargestGroup(11), 2);
  ASSERT_EQ(s.countLargestGroup(12), 3);
  ASSERT_EQ(s.countLargestGroup(13), 4);
  ASSERT_EQ(s.countLargestGroup(18), 9);
  ASSERT_EQ(s.countLargestGroup(19), 9);
  ASSERT_EQ(s.countLargestGroup(20), 1);
  ASSERT_EQ(s.countLargestGroup(21), 2);
  ASSERT_EQ(s.countLargestGroup(22), 3);
  ASSERT_EQ(s.countLargestGroup(27), 8);
  ASSERT_EQ(s.countLargestGroup(28), 8);
  ASSERT_EQ(s.countLargestGroup(29), 8);
  ASSERT_EQ(s.countLargestGroup(30), 1);
  return 0;
}

File: 1399.count-largest-group.rs
/*
 * @lc app=leetcode id=1399 lang=rust
 *
 * [1399] Count Largest Group
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn count_largest_group(n: i32) -> i32 {
        let mut cnt = [0; 37];
        let mut mx = 0;
        for i in 1..=n {
            let d = Solution::dsum(i) as usize;
            cnt[d] += 1;
            mx = mx.max(cnt[d]);
        }
        cnt.iter().filter(|&a| *a == mx).count() as i32
    }

    fn dsum(n: i32) -> i32 {
        if n == 0 {
            0
        } else {
            n % 10 + Solution::dsum(n / 10)
        }
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::count_largest_group(13), 4);
}

#[test]
fn test2() {
    assert_eq!(Solution::count_largest_group(2), 2);
}
